proposals/0000-extension-lifecycle-framework.md

```diff
@@ -0,0 +1,128 @@
+
+## 2. Proposed Change Specification
+
+### 2.1. Categories and Lifecycle
+Language extensions will be classified into the following categories
+ * `Experimental` extensions are undergoing active development. The syntax and semantics that are enabled by the extension are likely to change regularly. It is expected that most new language extensions will begin as experimental. At the time of writing, `LinearTypes` and `OverloadedRecordUpdate` seem to be in this category.
+ 
+ * `Mature` extensions are considered to be finished and are not expected to undergo regular changes. These features can be used without worry of unexpected changes, and they are not known to contain serious design or implementation deficiencies. Any breaking change to a mature extension will be announced well in advance of the change being made, with a migration path provided if possible. Some extensions in this category might be `MultiWayIf`, `MonoLocalBinds`, and `ViewPatterns`.
+
+ * `Deprecated` extensions are considered to be design or implementation dead ends, and should probably not be used in new code. Deprecating an extension _must_ come with a statement about the longevity of the extension. This need not be a commitment to remove it at a concrete time, but is a commitment to either remove it or transition it to another state at a time stated by GHC HQ when it is marked as `Deprecated`. Any extension will be deprecated prior to removal. Deprecated extensions are expected to include `IncoherentInstances`, `OverlappingInstances`, `Rank2Types`, `RecordPuns`, and `TypeInType`.
+ 
+ * `Legacy` extensions are explicitly not recommended for new code, much like `Deprecated`, however, they are expected to be supported indefinitely. This may be to maintain an older language set, maintain a bridge to a newer feature, or otherwise keep backward compatibility. Expected legacy extensions include `CUSKs`, `DeepSubsumption`,  and `NPlusKPatterns`.
+
+Above we list expected categorizations of several extensions based on statements in the user's guide. We do not attempt to prescribe any particular categorizations as part of this proposal and simply provide a small number of expectations from the perspective of someone reading the existing documentation.
+
+The expected extension lifecycle includes the following transitions:
+ * `Experimental` -> `Mature`
+ * `Experimental` -> `Deprecated`
+ * `Mature` -> `Legacy`
+ * `Mature` -> `Deprecated`
+ * `Legacy` -> `Deprecated`
+ * `Deprecated` -> `Legacy`
```

Comment:
Given that we're trying to describe the "happy path" here, the following transitions seem logical to me:

```suggestion
 * `Experimental` -> `Mature`
 * `Experimental` -> `Legacy`
 * `Experimental` -> `Deprecated`
 * `Mature` -> `Legacy`
 * `Legacy` -> `Deprecated`
```

An experimental extension might move to any state. Mature extensions might fall out of favour but should typically not be deprecated/removed (at least without lingering in the legacy state for a good while).

Of course we might change course e.g. a deprecated extension might move back to legacy in the light of user feedback. But that seems adequately covered by the next paragraph?

Reply:
I think this is all addressed in the current draft. Thanks!